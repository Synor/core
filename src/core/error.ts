/* eslint-disable no-dupe-class-members */

type MigrationRecord = import('./migration').MigrationRecord

type SynorDatabaseErrorType = 'lock_error' | 'unlock_error'
type SynorMigrationErrorType = 'not_found'
type SynorValidationErrorType = 'dirty' | 'hash_mismatch'

type SynorErrorType =
  | SynorDatabaseErrorType
  | SynorMigrationErrorType
  | SynorValidationErrorType
  | 'EXCEPTION'

export class SynorError extends Error {
  type: SynorErrorType
  meta: any

  constructor(
    message: string,
    meta: any = {},
    type: SynorErrorType = 'EXCEPTION'
  ) {
    super(message)

    Error.captureStackTrace(this, this.constructor)

    this.name = this.constructor.name
    this.type = type
    this.meta = meta
  }
}

type SynorDatabaseErrorMeta = {
  lockId?: string
}

export class SynorDatabaseError extends SynorError {
  type!: SynorDatabaseErrorType
  meta!: SynorDatabaseErrorMeta

  constructor(
    type: 'lock_error' | 'unlock_error',
    meta: Required<Pick<SynorDatabaseErrorMeta, 'lockId'>>
  )

  constructor(type: SynorDatabaseErrorType, meta: SynorDatabaseErrorMeta) {
    super('Database Error', meta, type)
  }
}

export class SynorMigrationError extends SynorError {
  type!: SynorMigrationErrorType
  meta!: Partial<MigrationRecord>

  constructor(type: SynorMigrationErrorType, meta: Partial<MigrationRecord>) {
    super('Migration Error', meta, type)
  }
}

export class SynorValidationError extends SynorError {
  type!: SynorValidationErrorType
  meta!: MigrationRecord

  constructor(type: SynorValidationErrorType, meta: MigrationRecord) {
    super('Validation Error', meta, type)
  }
}
