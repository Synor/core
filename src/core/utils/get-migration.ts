import { SynorMigration } from '../migration'

type MigrationSource = import('../migration').MigrationSource
type MigrationType = import('../migration').MigrationType
type MigrationVersion = import('../migration').MigrationVersion
type SourceEngine = import('../source').SourceEngine

export async function getMigration(
  source: SourceEngine,
  version: MigrationVersion,
  type: MigrationType
): Promise<MigrationSource | null> {
  const info = await source.get(version, type)

  if (!info) {
    return null
  }

  const content = await source.read(info)

  const migration = SynorMigration(info, content)

  return migration
}
