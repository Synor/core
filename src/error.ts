/* eslint-disable no-dupe-class-members */

type MigrationRecord = import('./migration').MigrationRecord

type SynorMigrationErrorType = 'not_found'
type SynorValidationErrorType = 'dirty' | 'hash_mismatch'

type SynorErrorType =
  | SynorMigrationErrorType
  | SynorValidationErrorType
  | 'invalid_filename'
  | 'exception'

type SynorErrorDataType = {
  dirty: MigrationRecord
  hash_mismatch: MigrationRecord
  not_found: Partial<MigrationRecord>
  invalid_filename: { filename: string }
  exception: Error | Record<string, any>
}

export class SynorError<
  ErrorType extends SynorErrorType = 'exception'
> extends Error {
  type: ErrorType
  data: SynorErrorDataType[ErrorType]

  /**
   * Creates a general instance of SynorError
   *
   * @param message Error message
   */
  constructor(message: string)

  /**
   * Creates a special instance of SynorError with `type` and `data`
   *
   * @param message Error message
   * @param type Error type
   * @param data Error data
   */
  constructor(
    message: string,
    type: ErrorType,
    data: SynorErrorDataType[ErrorType]
  )

  constructor(
    message: string,
    type: SynorErrorType = 'exception',
    data: SynorErrorDataType[SynorErrorType] = {}
  ) {
    super(message)

    Error.captureStackTrace(this, this.constructor)

    if (data instanceof Error) {
      this.stack = `${data.stack}\n${this.stack}`
    }

    this.name = this.constructor.name
    this.type = type as ErrorType
    this.data = data as SynorErrorDataType[ErrorType]
  }
}

export function toSynorError<T extends SynorErrorType = 'exception'>(
  error: Error | SynorError<T>
): SynorError<T> {
  if (error instanceof SynorError) {
    return error
  }

  return new SynorError(
    `Exception: ${error.message}`,
    'exception',
    error
  ) as SynorError<T>
}
