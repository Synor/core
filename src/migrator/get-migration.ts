import { SynorMigration } from '../migration'

type MigrationSource = import('../migration').MigrationSource
type MigrationType = import('../migration').MigrationType
type SourceEngine = import('../source').SourceEngine

export async function getMigration(
  source: SourceEngine,
  version: string,
  type: MigrationType
): Promise<MigrationSource | null> {
  const info = await source.get(version, type)

  if (!info) {
    return null
  }

  const content = await source.read(info)

  const migration = SynorMigration(info, content)

  return migration
}
