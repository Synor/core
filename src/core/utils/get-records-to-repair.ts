import { getMigration } from './get-migration'

type MigrationHistory = import('../migration').MigrationHistory
type SourceEngine = import('../source').SourceEngine

type MismatchedRecord = Pick<MigrationHistory[number], 'id' | 'hash'>

type MigrationVersion = MigrationHistory[number]['version']

export async function getRecordsToRepair(
  source: SourceEngine,
  baseVersion: MigrationVersion,
  history: MigrationHistory
): Promise<MismatchedRecord[]> {
  const appliedRecords = history.filter(
    ({ state, dirty }) => state === 'applied' && !dirty
  )

  const mismatchedRecords: MismatchedRecord[] = []

  for (const { id, version, type, hash } of appliedRecords) {
    if (version === baseVersion) {
      continue
    }

    const migration = await getMigration(source, version, type)

    if (!migration) {
      throw new Error(`NOT_FOUND: Migration(${version}.${type})`)
    }

    if (migration.hash !== hash) {
      mismatchedRecords.push({ id, hash: migration.hash })
    }
  }

  return mismatchedRecords
}
