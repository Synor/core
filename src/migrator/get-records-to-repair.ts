import { SynorError } from '../error'
import { getMigration } from './get-migration'

type MigrationHistory = import('../migration').MigrationHistory
type SourceEngine = import('../source').SourceEngine

type MismatchedRecord = Pick<MigrationHistory[number], 'id' | 'hash'>

export async function getRecordsToRepair(
  source: SourceEngine,
  baseVersion: string,
  history: MigrationHistory
): Promise<MismatchedRecord[]> {
  const appliedRecords = history.filter(
    ({ state, dirty }) => state === 'applied' && !dirty
  )

  const mismatchedRecords: MismatchedRecord[] = []

  for (const { id, version, type, title, hash } of appliedRecords) {
    if (version === baseVersion) {
      continue
    }

    const migration = await getMigration(source, version, type)

    if (!migration) {
      throw new SynorError(
        `Missing Migration Source => Version(${version}) Type(${type}) Title(${title})`,
        'not_found',
        { id, version, type, title }
      )
    }

    if (migration.hash !== hash) {
      mismatchedRecords.push({ id, hash: migration.hash })
    }
  }

  return mismatchedRecords
}
