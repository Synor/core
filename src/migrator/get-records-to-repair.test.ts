import { SynorError } from '../error'
import { getMigrationRecordInfo, getMigrationSource } from '../__utils__.test'
import * as getMigrationModule from './get-migration'
import { getRecordsToRepair } from './get-records-to-repair'

jest.mock('./get-migration', () => ({
  getMigration: jest.fn()
}))

describe('migrator:getRecordsToRepair', () => {
  const source = {}

  test('throws is MigrationSource not found', async () => {
    jest.spyOn(getMigrationModule, 'getMigration').mockReset()
    jest
      .spyOn(getMigrationModule, 'getMigration')
      .mockImplementationOnce(() => Promise.resolve(null))

    const recordInfos = [
      getMigrationRecordInfo({ id: 1, version: '01', type: 'do' })
    ]

    try {
      await getRecordsToRepair(source as any, '0', recordInfos)
    } catch (error) {
      expect(error).toBeInstanceOf(SynorError)
      expect(error.type).toMatchInlineSnapshot(`"not_found"`)
    }
  })

  test('skips dirty records', async () => {
    jest.spyOn(getMigrationModule, 'getMigration').mockReset()
    jest
      .spyOn(getMigrationModule, 'getMigration')
      .mockImplementationOnce(() => Promise.resolve(null))

    const recordInfos = [
      getMigrationRecordInfo({ id: 1, version: '01', type: 'do', dirty: true })
    ]

    await expect(
      getRecordsToRepair(source as any, '0', recordInfos)
    ).resolves.toMatchSnapshot()
  })

  test('skips record with base version', async () => {
    jest.spyOn(getMigrationModule, 'getMigration').mockReset()
    jest
      .spyOn(getMigrationModule, 'getMigration')
      .mockImplementationOnce(() => Promise.resolve(null))

    const recordInfos = [
      getMigrationRecordInfo({ id: 1, version: '01', type: 'do' })
    ]

    await expect(
      getRecordsToRepair(source as any, '01', recordInfos)
    ).resolves.toMatchSnapshot()
  })

  test('works as expected', async () => {
    jest.spyOn(getMigrationModule, 'getMigration').mockReset()
    jest
      .spyOn(getMigrationModule, 'getMigration')
      .mockResolvedValueOnce(getMigrationSource({ version: '01', hash: '1' }))
      .mockResolvedValueOnce(
        getMigrationSource({ version: '02', type: 'undo' })
      )
      .mockResolvedValueOnce(getMigrationSource({ version: '02', hash: '2.x' }))

    const recordInfos = [
      getMigrationRecordInfo({ id: 1, version: '01', type: 'do', hash: '1' }),
      getMigrationRecordInfo({
        id: 2,
        version: '02',
        type: 'do',
        hash: '2',
        state: 'reverted',
        revertedBy: 4
      }),
      getMigrationRecordInfo({
        id: 3,
        version: '03',
        type: 'do',
        hash: '3',
        state: 'reverted',
        revertedBy: 4
      }),
      getMigrationRecordInfo({ id: 4, version: '02', type: 'undo' }),
      getMigrationRecordInfo({ id: 5, version: '02', type: 'do', hash: '2' })
    ]

    await expect(
      getRecordsToRepair(source as any, '0', recordInfos)
    ).resolves.toMatchSnapshot()
  })
})
