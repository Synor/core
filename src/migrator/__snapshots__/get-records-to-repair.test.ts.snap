// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrator:getRecordsToRepair skips dirty records 1`] = `Array []`;

exports[`migrator:getRecordsToRepair skips record with base version 1`] = `Array []`;

exports[`migrator:getRecordsToRepair works as expected 1`] = `
Array [
  Object {
    "hash": "2.x",
    "id": 5,
  },
]
`;
